src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -443,7 +445,7 @@ private bool TryGetPropertyForInputModelProperty(Func<ObjectTypeProperty, bool>
             else
             {
                 //only load implementations for the base type
-                implementations = _derivedTypes!.Select(child => new ObjectTypeDiscriminatorImplementation(child.Name, _typeFactory.CreateType(child))).ToArray();
```

Comment:
We should use the `DiscriminatorValue`, not `Name`. Here is what we previously created:
```
case "Shark": return Shark.DeserializeShark(element);
case "Salmon": return Salmon.DeserializeSalmon(element);
```

That's wrong. It should be:
```
case "shark": return Shark.DeserializeShark(element);
case "salmon": return Salmon.DeserializeSalmon(element);
```